edb/ir/utils.py

```diff
@@ -290,3 +290,12 @@ def get_iterator_sets(stmt: irast.Stmt) -> Sequence[irast.Set]:
         iterators.extend(stmt.hoisted_iterators)
 
     return iterators
+
+
+def contains_dml(stmt: irast.Stmt) -> bool:
+    """Check whether a statement contains any DML in a subtree."""
+    # If this ends up being a perf problem, we can use a visitor
+    # directly and cache.
+    res = ast.find_children(stmt, lambda x: isinstance(x, irast.MutatingStmt),
```

Comment:
How often will this thing re-traverse the tree? Do we need to add a cache here?

Reply:
Once per FOR, so it is quadratic in the nesting depth of FOR statements, which didn't seem that likely to be bad to me.